### PR TITLE
Experiment 1 (changing q value)

### PR DESCRIPTION
**Pull request by zhouzypaul**

### CODE REVIEW

File: actgen/change_q.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        q_target[0][action] = toward_target
+
+        loss = torch.nn.functional.smooth_l1_loss(input=old_q_values, target=q_target)
+        loss.backward()
+        self.agent.optimizer.step()
+
+        self.agent.q.eval()
+        new_q_values = self.agent.q(state.float())
+
+        return old_q_values.detach().numpy()[0], new_q_values.detach().numpy()[0]
```
Comment (zhouzypaul) [inline actgen/change_q.py]: could you check that this is indeed correct/what we should we doing?

File: actgen/change_q.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        parser.add_argument('--out_file', type=str, default='results/change_q_metric.csv',
+                            help='Path to a output file to write to that will contain the computed metrics')
+        parser.add_argument('--test', default=False, action='store_true',
+                            help='Enable test mode for quickly checking configuration works')
+        parser.add_argument('--num_update', '-n', type=int, default=5,
+                            help='Number of times to update a particular action q value')
+        parser.add_argument('--delta_update', '-u', type=float, default=10.0,
+                            help='increase the q value by this much for every update applied')
+        parser.add_argument('--change_percentage_thresh', '-p', type=float, default=0.10,
+                            help='only changes past this percentage are considered when computing the metrics')
```
Comment (zhouzypaul) [inline actgen/change_q.py:53]: here is where I've stored the hyper params

File: actgen/change_q.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        loss = torch.nn.functional.smooth_l1_loss(input=old_q_values, target=q_target)
+        loss.backward()
+        self.agent.optimizer.step()
+
+        self.agent.q.eval()
+        new_q_values = self.agent.q(state.float())
+
+        return old_q_values.detach().numpy()[0], new_q_values.detach().numpy()[0]
+
+    def plot(self, old_values, new_values, action_idx):
```
Comment (camall3n) [inline actgen/change_q.py]: This could probably be a static method. It doesn't need `self`, since `old_values` and `new_values` are both length `n` already

File: actgen/change_q.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        # set up
+        trial = Trial(test=test)
+        s = trial.test_env.reset()  # one starting state
+        original_q_values = trial.original_q_values(s)  # q(s,a) for all a in actions
+
+        base_update_value = original_q_values[a] + delta_update
+        new_values = []
+        # update for a few consecutive times
+        for i in range(num_updates):
+            update_value = base_update_value + delta_update * i
```
Comment (camall3n) [inline actgen/change_q.py]: I also see that you're increasing the q-value delta with each iteration. This is probably fine, although it's unlikely that we'll reach the updated value in a single optimizer step, so we don't necessarily need to change the target each iteration. Here I'd opt for whatever results in the most concise explanation in the paper, which is probably to have a constant offset and take multiple steps toward the same target.

Comment (camall3n) [reply actgen/change_q.py]: The delta plots suggest that we are only moving by about 0.1, anyway, despite the first q-value target being 10.0 more than the starting point. So I doubt increasing the target each time has much of an effect.

File: actgen/change_q.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        # plot reference line at 0
+        plt.plot(possible_actions, np.zeros_like(possible_actions), '--k')
+
+        # plot all new values
+        for i, new_val in enumerate(new_values):
+            plt.plot(possible_actions, new_val - old_values, label='iter {}'.format(i))
+
+        plt.title('action {}'.format(action_idx))
+        plt.xlabel('{} different actions'.format(self.test_env.action_space.n))
+        plt.ylabel('q(s,a)')
```
Comment (camall3n) [inline actgen/change_q.py]: You can add a delta symbol to the y-label with `$\Delta$`. You may need to r-quote the string: `r'$\Delta$q(s,a)'`

File: requirements.txt [hunk 6]
```diff
@@ -12,8 +15,10 @@ py==1.10.0
 pyglet==1.5.0
 pyparsing==2.4.7
 pytest==6.2.1
+python-dateutil==2.8.1
 scipy==1.6.0
 seeding==0.1.0
+six==1.15.0
 toml==0.10.2
 torch==1.7.1
 tqdm==4.55.1
```
Comment (camall3n) [inline requirements.txt:24]: Are these updates for matplotlib?

**Review by camall3n — COMMENTED** [review]
See discussion on slack about SGD vs Adam

File: requirements.txt (hunk 6 above)
Comment (zhouzypaul) [reply requirements.txt:24]: yea!

File: actgen/change_q.py (hunk 4 above)
Comment (zhouzypaul) [reply actgen/change_q.py]: good point, the update is indeed happening very slow. thanks for the suggestion

File: actgen/change_q.py (hunk 3 above)
Comment (zhouzypaul) [reply actgen/change_q.py]: ah good point!

**Review by camall3n — APPROVED** [review]
I'm not sure if it makes sense to reuse the Trial class infrastructure for evaluate_dqn and change_q. Or if it does make sense, whether we should refactor it so there's a common Trial class instead of 3 different ones. In any case, this code works a lot better than nothing, so I'm approving the PR.

**Comment by camall3n** [comment]
I was able to reproduce this, so I'm merging it in
